helix-core/src/main/java/org/apache/helix/controller/stages/PerReplicaThrottleStage.java

```diff
@@ -220,6 +271,224 @@ private MessageOutput throttlePerReplicaMessages(IdealState idealState,
     return output;
   }
 
+  private void propagateCountsTopDown(StateModelDefinition stateModelDef,
```

Comment:
Let me give an example to illustrate

N1(O), N2(S), N3(O), message N3(O->S) should be classified as recovery.

Assuming without the propagateCounts logic,

Here, the expectedStateCountMap is {M->1, S->1, O->0}, given MIN_ACTIVE=2. 
The currentStateCountMap is {M->0, S->1, O->0}

See the message N3(O->S), the logic to test load/recovery is following:

```
if (isUpward && (currentCount < expectedCount)) {
          recoveryMessages.add(msg);
          currentStateCounts.put(toState, currentCount + 1);
        } else {
          loadMessages.add(msg);
        }
```
Here the currentCount (slave) is 1. The expectedCount is also 1.  Since current state no less than expected count for this state (SLAVE), it would classify the message as load. This is wrong of course.

With population logic, we have expectedStateCountMap is {M->1, S->2 O->3}. currentStateCountMap is {M->0, S->1, O->1}. This is going to work.

In theory,

When determining a message as LOAD or RECOVERY,  we look at the toState of this message. If the toState current count meet the required expected count of the toState, we will treat it as Load, otherwise, it is Recovery.

Reply:
This is a nice design, but does it cover this case, current state {M>0, S>2, 0>1}, and the message is O->S. The expectation is that the message will be done and then the S->M is the next one. However, it will be treated as loadbalance, right?

It seems to be a very difficult problem. But I think we need to address it because the original algorithm will generate the right result.

Reply:
Another potential flaw of this design is that we do not consider if a state transition is possible between 2 states. The current implementation implies that a lower priority state can always be transited to a higher priority state. This is not always true.

Reply:
As discussed offline, here is reasoning about first case as agreed.

The first issue raised can be illustrated use this example:
```
Preference list N0, N1, N2
Current N0 - O, N1 - S, N2 - S
current count  M(0), S(2), O (2)
Message
      N0 O -> S (Load)

```
The N0 (0->S) is classified as load in this case. Since eventually N0 will be master based on PreferenceList. So should we give it preference as recovery? 

For now, we decide not do this kind of optimization for the reason that classifying it as load won't delay a master being promoted to N1. Thus, eventually N0 would be promoted to Master once N0 arrive at slave state. The gist is best possible would step-by-step lead to preference list. For now, we can rely on best possible stage to make the choice. We will note this potential optimization in design doc.

Reply:
> Another potential flaw of this design is that we do not consider if a state transition is possible between 2 states. The current implementation implies that a lower priority state can always be transited to a higher priority state. This is not always true.

This is addressed with updated `propagateCountsTopDown` implementation as discussed. The higher priority state count would be added to all the lower states that can transit to the higher priority state.

Reply:
ping

Reply:
I think the new implementation reflects the overall design as we discussed. So I only have several comments about the code itself. But the overall direction looks fine to me.

But still, please try to schedule a meeting with Junkai, who might have a better idea.

Reply:
Ok, for now let me resolve this issue.